File: python/DiameterOfBTree/GreatestDistance.py

class Diameter(object):
    def __init__(self, left, distanceLeft, right, distanceRight, diameter):
        super(Diameter, self).__init__()
        self.left = left
        self.right = right
        self.distanceRight = distanceRight
        self.distanceLeft = distanceLeft
        self.diameter = diameter

    def getMaxDistance(self):
        return max(self.distanceRight, self.distanceLeft)

    def getEndmostNode(self):
        if self.distanceLeft > self.distanceRight:
            return self.left
        else:
            return self.right

    def increaseBy(self, tree):
        if self.distanceLeft > self.distanceRight:
            self.right = tree
            self.distanceRight = 0
            self.distanceLeft += 1
        else:
            self.left = tree
            self.distanceLeft = 0
            self.distanceRight += 1

        self.diameter += 1

    def __str__(self):
        return "diameter length: {}, left: {},  right: {}".format(
            self.diameter, self.left.key, self.right.key
        )


def findDiameter(tree):
    if tree.left:
        leftSub = findDiameter(tree.left)
    if tree.right:
        rightSub = findDiameter(tree.right)

    if tree.left and tree.right:
        diameter = max(leftSub.diameter, rightSub.diameter)
        newPath = leftSub.getMaxDistance() + rightSub.getMaxDistance() + 2
        if newPath > diameter:
            r = Diameter(
                leftSub.getEndmostNode(),
                leftSub.getMaxDistance() + 1,
                rightSub.getEndmostNode(),
                rightSub.getMaxDistance() + 1,
                newPath
            )
        else:
            r = leftSub.diameter \
                if leftSub.diameter > rightSub.diameter else rightSub

    elif tree.left or tree.right:
        r = leftSub if tree.left else rightSub
        if r.getMaxDistance() + 1 > r.diameter:
            r.increaseBy(tree)
        else:
            r.distanceLeft += 1
            r.distanceRight += 1
    else:
        r = Diameter(
            tree,
            0,
            tree,
            0,
            0
        )

    return r

File: java/DiameterOfBTree/src/muatik/GreatestDistance.java
package muatik;

public class GreatestDistance {
	Object left, right; // the farthest left/right node.
	int distanceLeft, distanceRight; // distance to the farthest left/right node.
	int diameter; // path from left to right
	
	public GreatestDistance(Object left, int distanceLeft, Object right, int distanceRight, int pathLength) {
		this.left = left;
		this.distanceLeft = distanceLeft;
		this.right = right;
		this.distanceRight = distanceRight;
		this.diameter = pathLength;
	}
	
	public Object getFarthestNode() {
		if (distanceLeft > distanceRight)
			return left;
		return right;
	}
	
	public int getMaxDistance() {
		if (distanceLeft > distanceRight)
			return distanceLeft;
		return distanceRight;
	}
	
	public void increaseBy(Object tree) {
		if (distanceLeft > distanceRight) {
			right = tree;
			distanceRight = 0;
			distanceLeft++;
		} else {
			left = tree;
			distanceLeft = 0;
			distanceRight++;
		}
		diameter++;
	}
	
	public static GreatestDistance get(BTree tree) {
		GreatestDistance leftSub  = null;
		GreatestDistance rightSub = null;
		GreatestDistance r = null;
		
		if (tree.left !=null) 
			leftSub = get(tree.left);
		
		if (tree.right !=null) 
			rightSub = get(tree.right);
		
		if (tree.left != null && tree.right != null) {
			
			if (
				(leftSub.getMaxDistance() + rightSub.getMaxDistance() + 2) 
					> Math.max(leftSub.diameter, rightSub.diameter)	
				) 
			{
				r = new GreatestDistance(
						leftSub.getFarthestNode(),
						leftSub.getMaxDistance() + 1,
						rightSub.getFarthestNode(), 
						rightSub.getMaxDistance() + 1,
						leftSub.getMaxDistance() + rightSub.getMaxDistance() + 2);
			} else {
				r = (leftSub.diameter > rightSub.diameter ? leftSub : rightSub);
			}
			
		}
		else if (tree.left != null || tree.right != null) {
			r = (tree.left == null? rightSub : leftSub);
			if (r.getMaxDistance() + 1 > r.diameter) {
				r.increaseBy(tree);
			} else {
				r.distanceLeft++;
				r.distanceRight++;
			}
		} else {
			// tree has neither left node nor right node. 
			r = new GreatestDistance(
					tree, // farthest left node
					0, // distance to the left node
					tree, // farthest right node
					0, // distance to the right node
					0);  // longest path's length
		}
		
		return r;
	}
}
